### DISCUSSION TITLE
Welcome to GraphEngine Discussions!

### DISCUSSION THREAD

**Discussion by shaobin** [microsoft/GraphEngine#343] [Announcements]
## Welcome!
We hope that you use Graph Engine Discussions to:
  * Ask questions you’re wondering about.
  * Share ideas that shape the future of Graph Engine. 
  * Engage with other community members.

  To get started, comment below : )

**Comment by TaviTruman** [comment]
@shaobin Thanks for starting this Dr. Shao. I hope folks will come here and share. Graph Engine is an amazing technology with so much incredible potential.

**Comment by TaviTruman** [comment]
Server/Proxy/Module - I'm working to understand how best to use Proxy. Can you share some typical use-cases around when best to use the Proxy model?

**Reply by shaobin** [reply]
The biggest (and the only) difference between a GE server and GE proxy is that: a server owns a data partition, a proxy does not. A typical user scenario is that we use proxies as query **aggregators**. Here is an example,

In the distributed setting, the data is partitioned and stored on many machines and a query needs to be executed on many machines. In this case, the proxy can serve as a middle layer between a GE client and the GE servers: the proxy takes the query submitted by the client, splits it into sub-queries that can be executed on different machines, dispatches the sub-queries to the machines, then collects the results of the sub-queries, **aggregates** the results of sub-queries, and sends the final result to the client.

**Reply by TaviTruman** [reply]
@shaobin Perfect! I think that makes the GE Proxy a fantastic candidate for a type of decentralized query agent.

**Comment by TaviTruman** [comment]
How does GE Availability Clustering compare to other horizontal cloud scalability models like Azure Service Fabric, Dapr, or even Docker?

**Reply by TaviTruman** [reply]
For those interested, over the past 6 years, all my development projects have been and are based on GE. I have learned a lot through detailed code sifting, tracing, and debugging of how the GE is put together. It is awesome. One of the areas that I have found to be very productive is the GE Availability Group capability. I have tried other Microservice containers and tried to use GE in Docker, Kubernetes, and Service Fabric, and none of them work as well as the native capability built into GE. I don't think I have exhaustive knowledge and know-how in other Microservice container environments, so there is space for others to examine, research and experiment. I will write about and share my learning when I have time and space.
